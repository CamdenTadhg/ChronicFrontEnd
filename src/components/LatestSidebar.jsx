import React, {useState, useEffect} from 'react';
import {Link} from 'react-router-dom';
import {useDispatch, useSelector} from 'react-redux';
import {selectProfile, selectLatest} from '../redux/selectors/selectors';
import ChronicAPI from '../api/chronicAPI';
import {fetchLatest} from '../redux/thunks/latestThunks';

function LatestSidebar() {
    const dispatch = useDispatch();
    const profile = useSelector(selectProfile);
    const latest = useSelector(selectLatest);
    const [articles, setArticles] = useState([]);
    const [loadingMessage, setLoadingMessage] = useState('Waking up server, please wait');
    const [showError, setShowError] = useState(false);

    const buildKeywords = (keywordsSource) => {
        let keywords = [];
        for (let diagnosis of keywordsSource){
            keywords.push(diagnosis.diagnosis);
            if (diagnosis.keywords){
                keywords = keywords.concat(diagnosis.keywords)
            }
        }
        return keywords;
    }
    
    useEffect(async () => {
        //if no user logged in, pull article ids on chronic illnesses in general
        if (!profile.userId){
            dispatch(fetchLatest());
        //else pull article ids on the user's chosen diagnoses & keywords
        } else {
            const keywords = buildKeywords(profile.diagnoses);
            dispatch(fetchLatest(keywords))
        }
    }, [profile]);

    useEffect(() => {
        if (!latest.loading){
            //then pull article details based on the articleIds in the store
            try {
                if (Array.isArray(latest.articleIds) && latest.articleIds.length === 0 && !latest.error){
                    setArticles([{PMID: 1, title: 'No articles found'}]);
                } else {
                    const sixArticleIds = latest.articleIds.slice(0, 7);
                    const getArticles = async () => {
                        try{
                            const articleDetails = await ChronicAPI.getArticles(sixArticleIds);
                            setArticles(articleDetails);
                        } catch (error) {
                            const articleDetails = [{PMID: null, title: error.message}];
                            setArticles(articleDetails);
                        }
                    }
                    getArticles();
                }
            } catch (error) {
            }
        }
    }, [latest.articleIds]);

    //give articles time to populate before displaying error message
    useEffect(() => {
        if (!latest.loading) {
            const timeoutId = setTimeout(() => {
                if (articles.length === 0){
                    setShowError(true);
                }
            }, 1000);
            return () => clearTimeout(timeoutId);
        }
    }, [latest.loading]);

    return (
        <>
            <Link to='/latest'><h2 className='latest-head'>The Latest</h2></Link>
            <div className='articles-sidebar'>
                {Array.isArray(articles) && articles.length > 0 ? 
                (articles.map((article) => (
                    <p key={article.PMID}>
                        <a href={`https://pubmed.ncbi.nlm.nih.gov/${article.PMID}`} target="_blank" rel="noopener noreferrer">{article.title}</a>
                    </p>
                )))
                : latest.loading === true || showError === false ? <p>{loadingMessage}</p>:
                <p>An error occurred. Please try again later</p>}
            </div>
        </>
    )
}

export default LatestSidebar;
