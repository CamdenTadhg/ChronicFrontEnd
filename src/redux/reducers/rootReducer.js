import {combineReducers} from 'redux';
import profileReducer from './profileReducer';
import trackingReducer from './trackingReducer';
import dataReducer from './dataReducer';
import latestReducer from './latestReducer';

const rootReducer = combineReducers({
    profile: profileReducer,
    tracking: trackingReducer,
    data: dataReducer,
    latest: latestReducer
});

export default rootReducer;