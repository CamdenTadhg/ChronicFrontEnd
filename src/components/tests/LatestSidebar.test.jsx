import React from 'react';
import {render, waitFor, fireEvent} from '@testing-library/react';
import {it, expect, vi, afterEach} from 'vitest';
import '@testing-library/jest-dom';
import {MemoryRouter, BrowserRouter} from 'react-router-dom';
import {configureStore} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';
import LatestSidebar from '../LatestSidebar';
import rootReducer from '../../redux/reducers/rootReducer';
import ChronicAPI from '../../api/chronicAPI';

afterEach(() => {
    vi.resetAllMocks();
})

it('renders without errors', async () => {
    const store = configureStore({reducer: rootReducer});
    ChronicAPI.getArticleIds = vi.fn();
    ChronicAPI.getArticleIds.mockResolvedValue([1234,2233]);
    ChronicAPI.getArticles = vi.fn();
    ChronicAPI.getArticles.mockResolvedValue([
        {PMID: 2345, title: 'a test article', abstract: "this is a test article"},
        {PMID: 5732, title: 'a second test article', abstract: "this is also a test article"}
    ]);
    await waitFor(() => {
        render(
            <Provider store={store}>
                <MemoryRouter>
                    <LatestSidebar/>
                </MemoryRouter>
            </Provider>
        )
    });
});

it('matches the snapshot', async () => {
    const store = configureStore({reducer: rootReducer});
    ChronicAPI.getArticleIds = vi.fn();
    ChronicAPI.getArticleIds.mockResolvedValue([1234,2233]);
    ChronicAPI.getArticles = vi.fn();
    ChronicAPI.getArticles.mockResolvedValue([
        {PMID: 2345, title: 'a test article', abstract: 'this is a test article'},
        {PMID: 5732, title: 'a second test article', abstract: 'this is also a test article'}
    ]);
    const latestSidebar = render(
        <Provider store={store}>
            <MemoryRouter>
                <LatestSidebar/>
            </MemoryRouter>
        </Provider>
    );
    await waitFor(() => expect(latestSidebar).toMatchSnapshot());
});

it('displays the appropriate content', async () => {
    const store = configureStore({reducer: rootReducer});
    ChronicAPI.getArticleIds = vi.fn();
    ChronicAPI.getArticleIds.mockResolvedValue([1234,2233]);
    ChronicAPI.getArticles = vi.fn();
    ChronicAPI.getArticles.mockResolvedValue([
        {PMID: 2345, title: 'a test article', abstract: 'this is a test article'},
        {PMID: 5732, title: 'a second test article', abstract: 'this is also a test article'}
    ]);
    const {getByText} = render(
        <Provider store={store}>
            <MemoryRouter>
                <LatestSidebar/>
            </MemoryRouter>
        </Provider>
    );
    await waitFor(() => expect(getByText('The Latest')).toBeInTheDocument());
});

it('pulls general articles if no user is logged in', async () => {
    const store = configureStore({reducer: rootReducer});
    ChronicAPI.getArticleIds = vi.fn();
    ChronicAPI.getArticleIds.mockResolvedValue([1234,2233]);
    ChronicAPI.getArticles = vi.fn();
    ChronicAPI.getArticles.mockResolvedValue([
        {PMID: 2345, title: 'a test article', abstract: 'this is a test article'},
        {PMID: 5732, title: 'a second test article', abstract: 'this is also a test article'}
    ]);
    const {getByText} = render(
        <Provider store={store}>
            <MemoryRouter>
                <LatestSidebar/>
            </MemoryRouter>
        </Provider>
    );
    await waitFor(() => {
        expect(ChronicAPI.getArticleIds).toHaveBeenCalledOnce();
        expect(ChronicAPI.getArticleIds).toHaveBeenCalledWith(['chronic illness']);
        expect(ChronicAPI.getArticles).toHaveBeenCalledOnce();
        expect(ChronicAPI.getArticles).toHaveBeenCalledWith([1234,2233]);
        expect(getByText('a test article')).toBeInTheDocument()
    });
});

it('pulls specific articles if a user is logged in', async () => {
    const store = configureStore({reducer: rootReducer, preloadedState: {
        profile: {
            userId: 3,
            email: 'test@test.com', 
            name: 'Test User',
            isAdmin: false,
            lastLogin: '2024-11-01Z10:00:00.000Z',
            diagnoses: [
                {diagnosis: 'myalgic encephalitis', keywords: ['CFS', 'chronic fatigue syndrome']}
            ],
            symptoms: [],
            medications: [],
            loading: false,
            error: null,
            history: {}
        }
    }});
    ChronicAPI.getArticleIds = vi.fn();
    ChronicAPI.getArticleIds.mockResolvedValue([1234,2233]);
    ChronicAPI.getArticles = vi.fn();
    ChronicAPI.getArticles.mockResolvedValue([
        {PMID: 2345, title: 'a test ME article', abstract: 'this is a test article about ME'},
        {PMID: 5732, title: 'a second test ME article', abstract: 'this is also a test article about ME'}
    ]);
    const {getByText} = render(
        <Provider store={store}>
            <MemoryRouter>
                <LatestSidebar/>
            </MemoryRouter>
        </Provider>
    );
    await waitFor(() => {
        expect(ChronicAPI.getArticleIds).toHaveBeenCalledOnce();
        expect(ChronicAPI.getArticleIds).toHaveBeenCalledWith(['myalgic encephalitis', 'CFS', 'chronic fatigue syndrome']);
        expect(ChronicAPI.getArticles).toHaveBeenCalledOnce();
        expect(ChronicAPI.getArticles).toHaveBeenCalledWith([1234,2233]);
        expect(getByText('a test ME article')).toBeInTheDocument();
    });
});

it('displays an appropriate message if no articles are found', async () => {
    const store = configureStore({reducer:rootReducer, preloadedState: {
        latest: {
            articleIds: [],
            loading: true,
            error: null
        }
    }});
    ChronicAPI.getArticleIds = vi.fn();
    ChronicAPI.getArticleIds.mockResolvedValue([]);
    ChronicAPI.getArticles = vi.fn();
    const {getByText} = render(
        <Provider store={store}>
            <MemoryRouter>
                <LatestSidebar/>
            </MemoryRouter>
        </Provider>
    );
    await waitFor(() => {
        expect(ChronicAPI.getArticleIds).toHaveBeenCalledOnce();
        expect(ChronicAPI.getArticles).not.toHaveBeenCalledOnce();
        expect(getByText('No articles found')).toBeInTheDocument()
    });
});

it('opens article links in a new tab',async () => {
    const store = configureStore({reducer: rootReducer});
    ChronicAPI.getArticleIds = vi.fn();
    ChronicAPI.getArticleIds.mockResolvedValue([1234,2233]);
    ChronicAPI.getArticles = vi.fn();
    ChronicAPI.getArticles.mockResolvedValue([
        {PMID: 2345, title: 'a test article', abstract: 'this is a test article'},
        {PMID: 5732, title: 'a second test article', abstract: 'this is also a test article'}
    ]);
    const {getByText} = render(
        <Provider store={store}>
            <MemoryRouter>
                <LatestSidebar/>
            </MemoryRouter>
        </Provider>
    );
    await waitFor(() => {
        fireEvent.click(getByText('a test article'));
        expect(getByText('The Latest')).toBeInTheDocument();
    });
});

it('navigates to the latest page on link click', async () => {
    const store = configureStore({reducer: rootReducer});
    const {getByText} = render(
        <Provider store={store}>
            <BrowserRouter>
                <LatestSidebar/>
            </BrowserRouter>
        </Provider>
    );
    fireEvent.click(getByText('The Latest'));
    await waitFor(() => expect(window.location.pathname).toBe('/latest'));
});