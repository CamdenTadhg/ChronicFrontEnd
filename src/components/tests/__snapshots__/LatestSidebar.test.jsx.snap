// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        href="/latest"
      >
        <h2
          class="latest-head"
        >
          The Latest
        </h2>
      </a>
      <div
        class="articles-sidebar"
      />
    </div>
  </body>,
  "container": <div>
    <a
      href="/latest"
    >
      <h2
        class="latest-head"
      >
        The Latest
      </h2>
    </a>
    <div
      class="articles-sidebar"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`matches the snapshot 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        href="/latest"
      >
        <h2
          class="latest-head"
        >
          The Latest
        </h2>
      </a>
      <div
        class="articles-sidebar"
      >
        <p>
          <a
            href="https://pubmed.ncbi.nlm.nih.gov/2345"
            rel="noopener noreferrer"
            target="_blank"
          >
            a test article
          </a>
        </p>
        <p>
          <a
            href="https://pubmed.ncbi.nlm.nih.gov/5732"
            rel="noopener noreferrer"
            target="_blank"
          >
            a second test article
          </a>
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <a
      href="/latest"
    >
      <h2
        class="latest-head"
      >
        The Latest
      </h2>
    </a>
    <div
      class="articles-sidebar"
    >
      <p>
        <a
          href="https://pubmed.ncbi.nlm.nih.gov/2345"
          rel="noopener noreferrer"
          target="_blank"
        >
          a test article
        </a>
      </p>
      <p>
        <a
          href="https://pubmed.ncbi.nlm.nih.gov/5732"
          rel="noopener noreferrer"
          target="_blank"
        >
          a second test article
        </a>
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`matches the snapshot 3`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        href="/latest"
      >
        <h2
          class="latest-head"
        >
          The Latest
        </h2>
      </a>
      <div
        class="articles-sidebar"
      >
        <p>
          <a
            href="https://pubmed.ncbi.nlm.nih.gov/1"
            rel="noopener noreferrer"
            target="_blank"
          >
            No articles found
          </a>
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <a
      href="/latest"
    >
      <h2
        class="latest-head"
      >
        The Latest
      </h2>
    </a>
    <div
      class="articles-sidebar"
    >
      <p>
        <a
          href="https://pubmed.ncbi.nlm.nih.gov/1"
          rel="noopener noreferrer"
          target="_blank"
        >
          No articles found
        </a>
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`matches the snapshot 4`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        href="/latest"
      >
        <h2
          class="latest-head"
        >
          The Latest
        </h2>
      </a>
      <div
        class="articles-sidebar"
      >
        <p>
          <a
            href="https://pubmed.ncbi.nlm.nih.gov/2345"
            rel="noopener noreferrer"
            target="_blank"
          >
            a test article
          </a>
        </p>
        <p>
          <a
            href="https://pubmed.ncbi.nlm.nih.gov/5732"
            rel="noopener noreferrer"
            target="_blank"
          >
            a second test article
          </a>
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <a
      href="/latest"
    >
      <h2
        class="latest-head"
      >
        The Latest
      </h2>
    </a>
    <div
      class="articles-sidebar"
    >
      <p>
        <a
          href="https://pubmed.ncbi.nlm.nih.gov/2345"
          rel="noopener noreferrer"
          target="_blank"
        >
          a test article
        </a>
      </p>
      <p>
        <a
          href="https://pubmed.ncbi.nlm.nih.gov/5732"
          rel="noopener noreferrer"
          target="_blank"
        >
          a second test article
        </a>
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
